### PR TITLE
crypto: Add support for non-IV-prefixed CBC ciphers in mbedTLS shim

### LINKED ISSUE #20561: Crypto API: Separate IV from ciphertext based on struct cipher_ctx::flags
**Problem:**
The crypto API now prefixes the IV to to the ciphertext. This is because Tinycrypt expects the IV and ciphertext to be in the contiguous memory for better performance. This is a problem if applications what to do inplace encrypt/decrypt (`in_buf == out_buf`) as now, the `in_buf` has to be larger by 2x block length and the IV has to be removed manually. Besides, not all use-cases expect IV to be prefixed to the ciphertext.

**Proposed Solution:**
Create new flag and pass it in struct `cipher_ctx::flags` that would allow `out_buf` to have only ciphertext. This means the TC shim would now have to do some extra buffer copy but MbedTLS can skip some.

I will raise a PR based on comments here.

### PR DESCRIPTION
**Pull request by sidcha**
This patch adds a new capability flag `CAP_NO_IV_PREFIX` that allows CBC operation without prefixing the IV to the output ciphertext.

When `CAP_NO_IV_PREFIX` is active, the IV passed to cipher_cbc_op() is not modified. So the length of the ciphertext block is equal to the plaintext block, allowing in-place encryption/decryption.

This flag is useful when the IV is derived from other sources (for instance a chained MAC serving as IV) and not directly transmitted to the other end.

**Notes:**
* Fixes: #20561
* Required for: #19483

### CODE REVIEW

File: drivers/crypto/crypto_mtls_shim.c
Comment (tbursztyka) [inline drivers/crypto/crypto_mtls_shim.c]: I cannot remember why iv is of 16bytes fixed length here, is it because of cbc or because of mbedtls? 

Could you detail? (note: I know next to nothing about crypto)

File: include/crypto/cipher_structs.h
Comment (tbursztyka) [inline include/crypto/cipher_structs.h]: you will need to put this api modification into its own commit. Then comes the mbedtls commit 

What about tc_shim commit as well? I wonder how you will be addressing ctr operator where (according to documentation) ivlen = keylen - ctrlen, so is of variable length.

File: drivers/crypto/crypto_mtls_shim.c
Comment (tbursztyka) [inline drivers/crypto/crypto_mtls_shim.c]: if you make a macro, make it in capital letters.

**Review by tbursztyka — DISMISSED** [review]

File: drivers/crypto/crypto_mtls_shim.c
Comment (cbsiddharth) [reply drivers/crypto/crypto_mtls_shim.c]: This is part of the other PR #20434. Will fix there.

Comment (tbursztyka) [reply drivers/crypto/crypto_mtls_shim.c]: sure

File: drivers/crypto/crypto_mtls_shim.c
Comment (cbsiddharth) [reply drivers/crypto/crypto_mtls_shim.c]: In AES-CBC, IV length is always 1 block size (128 bits) regardless of key lengths (AES-128, AES-256 etc.,) .

Comment (tbursztyka) [reply drivers/crypto/crypto_mtls_shim.c]: ok

File: include/crypto/cipher_structs.h
Comment (cbsiddharth) [reply include/crypto/cipher_structs.h]: > put this api modification into its own commit
OK.

>What about tc_shim
These changes are required for another PR I'm working. I'll get to tc_shim later. Anyway each shim has to advertise its capabilities separately. So I guess it shouldn't be an issue.

>how you will be addressing ctr operator where (according to documentation) ivlen = keylen - ctrlen, so is of variable length
mbdtls_shim currently doesn't support CTR mode. One quick look at the tc_shim suggests that IV isn't prefixed in CTR mode, but need to take a closer look to be sure.

Comment (cbsiddharth) [reply include/crypto/cipher_structs.h]: Done.

**Comment by cbsiddharth** [comment]
CI Failure looks unrelated. Can someone please verify?

**Comment by cbsiddharth** [comment]
@carlescufi, @nashif, @galak please review

**Review by d3zd3z — APPROVED** [review]

File: drivers/crypto/crypto_mtls_shim.c
Comment (cbsiddharth) [reply drivers/crypto/crypto_mtls_shim.c]: @tbursztyka Done. Please re-review.

**Comment by cbsiddharth** [comment]
Ping :)

**Comment by cbsiddharth** [comment]
Can someone please review/merge this?

**Comment by pfalcon** [comment]
> Can someone please review/merge this?

Trying to act on this request, being someone in the reviewer list, I find that it's somewhat unclear what/why is being done from just commit messages, and requires long tracing thru the linked issues to understand "why". Commit messages is what we'll have in the immediate project history, something we'll start to look at when we get a CVE, etc, etc. So, they rather be clear, especially for security-related patches.

So, if this patch fixes what is effectively a deficiency of Zephyr crypto API, where IV is conflated with input cleartext or output cyphertext, while the common baseline in cryptography is to treat those 3 as clearly distinct entities, then ... - I'd personally prefer the commit messages to say something just like that. What I see now at first glance is that it adds yet another mostly random feature of not very clear purpose, sorry.

**Comment by cbsiddharth** [comment]
> I find that it's somewhat unclear what/why is being done from just commit messages

@pfalcon, reading the commit message after so long, I realize how unclear it was. Sorry about that. I've reworded the commit message now, do let me know if this one is okay.

**Comment by pfalcon** [comment]
@cbsiddharth: Thanks for the updates. As I mentioned, I'm not familiar with Crypto API, so taking a chance to familiar myself and spot/act on any issues seen around it (e.g. https://github.com/zephyrproject-rtos/zephyr/issues/21820). Also, to clarify, I don't have merge rights, so there will be wait on someone with merge rights. From my side, I'm happy to work towards ensuring that crypto API gets into better shape than it was.

Typo in the new commit message:

> IV to be prefixed to the cipher/pain text for

Should be "plain".

**Comment by pfalcon** [comment]
And at the same line:

> cipher/pain text for performnace

Typo in "performance", if you care.

**Comment by cbsiddharth** [comment]
> Typo in the new commit message

Fixed. Thanks.

**Comment by pfalcon** [comment]
> Fixed. Thanks.

Cool, by I now lost my in-progress batched review comments ;-). This is first time this happens to me. An apparent danger of batching review comments vs posting them immediately. Nevermind of course, just a note to interested parties.

File: drivers/crypto/crypto_mtls_shim.c
Comment (pfalcon) [inline drivers/crypto/crypto_mtls_shim.c]: Can we preserve this comment, moving to the appropriate branch, perhaps extending as "Prefix IV to ciphertext, which is default behavior of Zephyr crypto API, unless CAP_NO_IV_PREFIX is requested."

File: drivers/crypto/crypto_mtls_shim.c
Comment (pfalcon) [inline drivers/crypto/crypto_mtls_shim.c]: This shows how weird is the old behavior: it allows to pass in arbitrary `iv` pointer, but still assumes there's an IV to skip in the input packet ;-/.

Comment (cbsiddharth) [reply drivers/crypto/crypto_mtls_shim.c]: I believe that this is embedtls shim's idiosyncrasy. Zephyr's crypto API didn't make any assumptions there.

File: drivers/crypto/crypto_mtls_shim.c
Comment (cbsiddharth) [reply drivers/crypto/crypto_mtls_shim.c]: Done.

Comment (pfalcon) [reply drivers/crypto/crypto_mtls_shim.c]: Thanks. Looking forward to a push to approve it then. Let's see if that random CI failure will go away too...

**Review by pfalcon — APPROVED** [review]
I vouch that I fully understood changes done here and they're good improvement of the Zephyr crypto API indeed. Thanks!

**Comment by pfalcon** [comment]
@cbsiddharth: Ok, hopefully with 2 vote-ups this will be merged quicker. Otherwise, I see that you got a few of your patches merged already, so you know merges do happen ;-). One way to speed it up might be to hang out on the Slack channel and ping folks with merge rights. Final way is to bring it up to the TSC meeting again, but that's usually only required for PRs with controversies, definitely not the case here.

**Comment by cbsiddharth** [comment]
> so you know merges do happen ;-)

Ohh, I do know they will, eventually. I'm just worried about an upcoming change that's slightly bigger than this one. Don't know how long that's going to take ;-).

**Comment by pfalcon** [comment]
>  I'm just worried about an upcoming change that's slightly bigger than this one. Don't know how long that's going to take ;-).

All I can say is that I personally appreciate submitting a small, clear, and standalone changes. In this case, winter holidays definitely affected timing. Big changes are indeed may take longer to process, and IMHO regularly merged by "leap of faith", with common cases of regressions/unexpected behavior detected months later. Oh, btw, do you know that we have tests support in Zephyr tree? ;-)

Good luck with your next patches anyway!

**Comment by pfalcon** [comment]
@cbsiddharth, So, https://github.com/zephyrproject-rtos/zephyr/pull/21827 was merged, and there's indeed a conflict now, please rebase.

This PR also fails CI check, and that would be another reason why it's not being merged. Actually, people oftentimes even skip reviewing such, apparently thinking "if it doesn't yet even build, then it's clearly not ready for review". Of course, there can be false positives, but dealing with those effectively falls on the PR author. This PR seems to be unlucky to trigger some random failures  with x86_64 arch. I suggest spending 5 mins to click thru Shippable UI to be able to triage these. One obvious way to restart the build is to rebase your PR. If that doesn't work, you can close and reopen the PR. If you see that some failure is seemingly unrelated, but systematic, you would need to call maintainers of that area for help (git log / CODEOWNERS file in the repo top dir).

**Comment by cbsiddharth** [comment]
@pfalcon, thank you for the tips. I'll rebase and sort out the build failure in the morning (I'm traveling now).

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
